drivers/spi/spi_rpi_pico_pio.c

```diff
@@ -0,0 +1,403 @@
+			/* Send 0 in the case of read only operation */
+			txrx = 0;
+
+			if (txbuf) {
+				txrx = ((uint8_t *)txbuf)[data->tx_count];
+			}
+			spi_pico_pio_sm_put8(data->pio, data->pio_sm, txrx);
+			data->tx_count++;
+			fifo_cnt++;
+		}
+
+		while ((!pio_sm_is_rx_fifo_empty(data->pio, data->pio_sm)) &&
+		       data->rx_count < chunk_len && fifo_cnt > 0) {
+			txrx = spi_pico_pio_sm_get8(data->pio, data->pio_sm);
+
+			/* Discard received data if rx buffer not assigned */
+			if (rxbuf) {
+				((uint8_t *)rxbuf)[data->rx_count] = (uint8_t)txrx;
+			}
+			data->rx_count++;
+			fifo_cnt--;
+		}
+	}
+}
+
+static int spi_pico_pio_transceive_impl(const struct device *dev, const struct spi_config *spi_cfg,
+					const struct spi_buf_set *tx_bufs,
+					const struct spi_buf_set *rx_bufs, bool asynchronous,
+					spi_callback_t cb, void *userdata)
+{
+	const struct spi_pico_pio_config *dev_cfg = dev->config;
+	struct spi_pico_pio_data *data = dev->data;
+	struct spi_context *spi_ctx = &data->spi_ctx;
+	int rc = 0;
+
+	spi_context_lock(spi_ctx, asynchronous, cb, userdata, spi_cfg);
+
+	rc = spi_pico_pio_configure(dev_cfg, data, spi_cfg);
+	if (rc < 0) {
+		goto error;
+	}
+
+	spi_context_buffers_setup(spi_ctx, tx_bufs, rx_bufs, data->dfs);
+	spi_context_cs_control(spi_ctx, true);
+
+	do {
+		spi_pico_pio_txrx(dev);
+		spi_context_update_tx(spi_ctx, 1, data->tx_count);
+		spi_context_update_rx(spi_ctx, 1, data->rx_count);
+	} while (spi_pico_pio_transfer_ongoing(data));
+
+	spi_context_cs_control(spi_ctx, false);
+
+error:
+	spi_context_release(spi_ctx, rc);
+
+	return rc;
+}
+
+static int spi_pico_pio_transceive(const struct device *dev, const struct spi_config *spi_cfg,
+				   const struct spi_buf_set *tx_bufs,
+				   const struct spi_buf_set *rx_bufs)
+{
+	return spi_pico_pio_transceive_impl(dev, spi_cfg, tx_bufs, rx_bufs, false, NULL, NULL);
+}
+
+int spi_pico_pio_release(const struct device *dev, const struct spi_config *spi_cfg)
+{
+	struct spi_pico_pio_data *data = dev->data;
+
+	spi_context_unlock_unconditionally(&data->spi_ctx);
+
+	return 0;
+}
+
+static struct spi_driver_api spi_pico_pio_api = {
+	.transceive = spi_pico_pio_transceive,
+	.release = spi_pico_pio_release,
+};
+
+static int config_gpio(const struct gpio_dt_spec *gpio, const char *tag, int mode)
+{
+	int rc = 0;
+
+	if (!device_is_ready(gpio->port)) {
+		LOG_ERR("GPIO port for %s pin is not ready", tag);
+		return -ENODEV;
+	}
+
+	rc = gpio_pin_configure_dt(gpio, mode);
+	if (rc < 0) {
+		LOG_ERR("Couldn't configure %s pin; (%d)", tag, rc);
+		return rc;
+	}
+
+	return 0;
+}
+
+int spi_pico_pio_init(const struct device *dev)
+{
+	const struct spi_pico_pio_config *dev_cfg = dev->config;
+	struct spi_pico_pio_data *data = dev->data;
+	int rc;
+
+	rc = pinctrl_apply_state(dev_cfg->pin_cfg, PINCTRL_STATE_DEFAULT);
+	if (rc) {
+		LOG_ERR("Failed to apply pinctrl state");
+		return rc;
+	}
+
+	rc = config_gpio(&dev_cfg->clk_gpio, "clk", GPIO_OUTPUT_ACTIVE);
```

Comment:
Why is this not handled by pinctrl?

Reply:
Turns out you're right - it IS handled by pinctrl.  That code is redundant;  I deleted it.